### ISSUE TITLE
"requires_grad" attribute on intermediate TensorProxies is unused and misleading

### ISSUE THREAD

**Issue by IvanYashchuk** [Lightning-AI/lightning-thunder#1570]
_Originally posted by @IvanYashchuk in https://github.com/Lightning-AI/lightning-thunder/pull/1563#discussion_r1890146557_

`requires_grad` of intermediate TensorProxies is ignored in our automatic differentiation code because we haven't done the work of properly threading this property through all computations.
We should remove the ability to query `.requires_grad` from intermediate TensorProxies completely to avoid similar bugs in the future. This can be achieved by introducing a separate "InputTensorProxy" which has this attribute and removing it from the regular TensorProxy.

**Comment by IvanYashchuk** [comment]
Remove it and restore it when it's correct. Today requires_grad is checked and used only for inputs of computation traces and ignored otherwise. Currently, the implemented logic is that every intermediate requires a gradient. The result of intermediate gradient calculations may be DCE'd if it remains unused and is not an output of the backward trace. The output of the backward trace is set to use `None` for every input tensor which doesn't require a gradient.

PyTorch has at least two ways of stopping gradient propagation:
1. setting `requires_grad=False` on an intermediate tensor
2. using `.detach()` on an intermediate tensor

We should support tracing both and a special operator to act as a gradient sink could be "detach". It should be possible to coerce the `tensor.requires_grad` assignment to False to `detach`. I don't think setting `tensor.requires_grad=True` does anything useful except for undoing the False assignment.

**Comment by t-vi** [comment]
We're not there yet, but there may be cases where we want to create new autograd leaves when we look beyond tracing models.
